gammapy/datasets/core.py

```diff
@@ -64,18 +74,23 @@ def stat_array(self):
         """Statistic array, one value per data point."""
 
     def copy(self, name=None):
-        """A deep copy."""
+        """A deep copy.
+
+        Parameters
+        ----------
+        name : str
+            Name of the copied dataset
+
+        Returns
+        -------
+        dataset : `Dataset`
+            Copied datasets.
+        """
         new = copy.deepcopy(self)
         name = make_name(name)
         new._name = name
-
-        # propagate new dataset name
-        if new._models is not None:
-            for m in new._models:
-                if m.datasets_names is not None:
-                    for k, d in enumerate(m.datasets_names):
-                        if d == self.name:
-                            m.datasets_names[k] = name
+        # TODO: check the model behaviour?
+        new.models = None
```

Comment:
I don't see a problem in keeping the models for .copy().

Reply:
I agree it's not necessarily a problem, however the way it was used in our code, was basically to copy and modify the model, which I think is non-ideal, because it makes the copy of the data for no good reason. I think in general there are not many use-cases, where a copy of the data is needed, so I think we should advertise `Models.copy()` instead. In the rare cases, where the data is copied, one can easily re-set the model and get's a uniform behaviour with all the other dataset methods, so I'd prefer to keep this change.

Reply:
If users have to re-set the models by themselfs, they will have also to propagate the new datasets names which is not very convenient (using append rather than replace as there is no copy).
Concerning uniformity with the other methods, maybe you can add an option keep_models : bool, that adapt or reset the models, and I will introduce that in a separate PR for the other methods (already started for .cutout())

Reply:
I'd be open to include something like this, but maybe we do things in order and start at the `Models` level first? Like:
- `Models.reassign(dataset_name=, dataset_name_new=)`
- `Models.select_region(region, margin="0.1 deg")`
- `Models.select_mask(mask, margin="0.1 deg")`
- Add `WcsGeom.footprint()`, which returns a region...

Then my proposal would be that we fix the `npred_background()` behaviour and only finally introduce the (optional) model handling. Does this sound reasonable?

Reply:
ok let's discuss that later